### ISSUE TITLE
Block Bootstrap

### ISSUE THREAD

**Issue by jonaslandsgesell** [scipy/scipy#14431]
**Is your feature request related to a problem? Please describe.** i.i.d Bootstrap does not work well for time series data which have a time correlation

**Describe the solution you'd like**
Block bootstrap addresses the problem of temporal correlation. It would be nice to have an option to apply block bootstrap in scipy.stats.bootstrap

**Describe alternatives you've considered**
This project implements sever block bootstrap methods
https://pypi.org/project/recombinator/

**Additional context (e.g. screenshots)**
Ping to the original authors of scipy.stats.bootstrap @mdhaber @tupui

**Comment by mdhaber** [comment]
Thanks for the suggestion @jonaslandsgesell . There are a few other things on my to-do list, but I'll be thinking about it.

**Comment by mdhaber** [comment]
@jonaslandsgesell If I'm understanding correctly, all that changes is the resampling procedure - instead of resampling individual observations, we resample blocks of observations until we have a new sample that is the same size as the original.

In that case, we can use the existing `bootstrap` function to do the job, but first we have to change the basis of our data and statistic, in a sense. Instead of the original data, we pass `bootstrap` _indices_ corresponding with our blocks (i.e. `np.arange(0, n_blocks)`, and instead of the original statistic function, we pass a statistic function that works on (resampled) indices of the blocks.

For example, the non-overlapping block bootstrap could be done to bootstrap the slope of some time series data like:
```python3
import numpy as np
from scipy.stats import linregress, bootstrap

# generate data
n = 300
np.random.seed(0)  # not supposed to use this anymore 
x = np.linspace(0, 10, n)
e = np.random.rand(len(x))
y = 3*x + 4 + np.sin(x/10*2*np.pi*20)*(1 + x/3) + e

## Estimate slope and intercept and separate trend from residuals
model = linregress(x, y)
trend = model.slope*x + model.intercept
residuals = y - trend

## Block bootstrap the intercept

l = 15  # block length
blocks = residuals.reshape(-1, l)

def my_statistic(i):
    y = blocks[i].ravel() + trend  # resample from block indices
    return linregress(x, y).slope

i = np.arange(n//l)  # original block indices
# BCa involves the jacknnife, and I'm not 100% sure how this should
# be handled, as it changes the number of observations in the resample.
# Using "basic" bootstrap instead.
res = bootstrap((i,), my_statistic, method='basic', vectorized=False)
```
(This brings to mind that it would also be nice to allow the statistic function to return several outputs and for `bootstrap` to return the bootstrap distribution so the user could visualize it like the plot near the end of https://blogs.sas.com/content/iml/2021/01/06/simple-block-bootstrap-sas.html.)

Other methods could be done similarly, essentially just changing the way that `blocks` is prepared and the definition of `i`. (It would be slightly inefficient without changes to `bootstrap` internals: `bootstrap` would need to resample from all the possible block choices, but you'd only _use_ the first few resampled `(n/l)` indices until you have enough data.)

This is very similar to how paired sample statistics are treated. If I were to implement this in `bootstrap`, I would treat it very similarly to the `paired` parameter by wrapping the user-provided statistic in a function that performs this sort of "change of basis". A little care needs to be taken when the samples are N-d arrays and when `vectorized=True`, but it's not so bad.

Reviewer attention is in short supply.  I have about a dozen PRs dating back to December that are waiting on review by other maintainers. Some of them are higher priority, IMO, like gh-13312 and gh-13899, and some are very easy, like gh-14207. So I would be happy to add this, but I would prefer it come after the higher-priority ones, and there would need to be someone lined up to take a look.

**Comment by jonaslandsgesell** [comment]
Hello @mdhaber the description you posted is also my understanding of the block bootstrap method.
Surely, I would be interested in giving some feedback on the implementation and the corresponding tests.

**Comment by mdhaber** [comment]
That would be terrific if you'd take a look. I'd also need another maintainer to take a look, though. So when those other PRs are set and there a maintainer interested here, I'd be happy to add this and I'll @mention you then.
